File: apps/fsharp/src/MachineLearning/Chapter14/KMeans.fs
module MachineLearning.Chapter14.KMeans

open MachineLearning.Chapter02.Random

/// 1 つの点。列の順に値を並べた配列
type Point = float[]

/// 2 点間のユークリッド距離の 2 乗
let squaredDistance (a: Point) (b: Point) : float =
    Array.map2 (fun x y -> (x - y) * (x - y)) a b |> Array.sum

/// 各点を、最も近い中心の番号（クラスタ番号）に割り当てる
let assignClusters (centers: Point[]) (points: Point[]) : int[] =
    points
    |> Array.map (fun point ->
        [| 0 .. centers.Length - 1 |]
        |> Array.minBy (fun k -> squaredDistance point centers[k]))

/// クラスタごとに、割り当てられた点の平均を新しい中心にする
let updateCenters (points: Point[]) (labels: int[]) (previousCenters: Point[]) : Point[] =
    previousCenters
    |> Array.mapi (fun k previous ->
        let members =
            Array.zip points labels
            |> Array.filter (fun (_, label) -> label = k)
            |> Array.map fst

        match members with
        | [||] -> previous
        | _ -> members |> Array.transpose |> Array.map Array.average)

/// クラスタリングの結果。Labels は点ごとのクラスタ番号
type KMeansResult =
    {
        Labels: int[]
        Centers: Point[]
        Sse: float
    }

/// SSE。各点と、その点が属するクラスタの中心との距離の 2 乗の合計
let sumOfSquaredErrors (points: Point[]) (labels: int[]) (centers: Point[]) : float =
    Array.map2 (fun point label -> squaredDistance point centers[label]) points labels
    |> Array.sum

/// 反復回数の上限の既定値
[<Literal>]
let MaxIterations = 300

/// 中心が変わらなくなるか、残りの反復回数が 0 になるまで、割り当てと中心の更新を繰り返す
[<TailCall>]
let rec private iterate (iterationsLeft: int) (points: Point[]) (centers: Point[]) : Point[] =
    let next = updateCenters points (assignClusters centers points) centers

    if next = centers || iterationsLeft = 1 then
        next
    else
        iterate (iterationsLeft - 1) points next

/// 反復回数の上限を指定して、初期中心から K-means でクラスタリングする
let kmeansWith (maxIterations: int) (initialCenters: Point[]) (points: Point[]) : KMeansResult =
    let centers = iterate maxIterations points initialCenters
    let labels = assignClusters centers points

    {
        Labels = labels
        Centers = centers
        Sse = sumOfSquaredErrors points labels centers
    }

/// 反復回数の上限を既定値にした K-means
let kmeans: Point[] -> Point[] -> KMeansResult = kmeansWith MaxIterations

/// 点の番号をシード付きで並べ替え、先頭の nClusters 個の点を初期中心にする
let chooseInitialCenters (nClusters: int) (seed: int) (points: Point[]) : Point[] =
    [ 0 .. points.Length - 1 ]
    |> shuffle seed
    |> List.truncate nClusters
    |> List.map (fun i -> points[i])
    |> List.toArray

/// 初期中心の候補ごとにクラスタリングし、SSE が最小の結果を返す
let bestKMeans (candidates: Point[] list) (points: Point[]) : KMeansResult =
    candidates
    |> List.map (fun initialCenters -> kmeans initialCenters points)
    |> List.minBy (fun result -> result.Sse)

/// シードを 1 ずつずらして初期中心を nInit 通り選び、SSE が最小の結果を返す
let kmeansWithRestarts (nInit: int) (nClusters: int) (seed: int) (points: Point[]) : KMeansResult =
    let candidates =
        [ for i in 0 .. nInit - 1 -> chooseInitialCenters nClusters (seed + i) points ]

    bestKMeans candidates points

/// クラスタ数ごとに、初期中心を nInit 通り試した最小の SSE を求める（エルボー法）
let sseByClusterCount (nInit: int) (seed: int) (clusterCounts: int list) (points: Point[]) : (int * float) list =
    clusterCounts
    |> List.map (fun n -> n, (kmeansWithRestarts nInit n seed points).Sse)

File: apps/csharp/src/MachineLearning/Chapter14/KMeans.cs
namespace MachineLearning.Chapter14;

using MachineLearning.Chapter02;
using Point = System.Collections.Generic.IReadOnlyList<double>;

/// <summary>クラスタリングの結果。</summary>
/// <param name="Labels">点ごとのクラスタ番号</param>
/// <param name="Centers">クラスタごとの中心</param>
/// <param name="Sse">各点と、その点が属するクラスタの中心との距離の 2 乗の合計</param>
public sealed record KMeansResult(IReadOnlyList<int> Labels, IReadOnlyList<Point> Centers, double Sse);

/// <summary>
/// 点を、最も近い中心のクラスタに割り当てることを繰り返す K-means。
/// 初期中心は引数で受け取り、選び方（ランダム・やり直し）は外側で決める。
/// </summary>
public static class KMeans
{
    /// <summary>反復回数の上限の既定値。</summary>
    public const int MaxIterations = 300;

    /// <summary>2 点間のユークリッド距離の 2 乗。大小を比べるだけなので平方根は取らない。</summary>
    public static double SquaredDistance(Point a, Point b)
    {
        ArgumentNullException.ThrowIfNull(a);
        ArgumentNullException.ThrowIfNull(b);
        var sum = 0.0;
        for (var i = 0; i < a.Count; i++)
        {
            sum += (a[i] - b[i]) * (a[i] - b[i]);
        }

        return sum;
    }

    /// <summary>各点を、最も近い中心の番号（クラスタ番号）に割り当てる。</summary>
    public static IReadOnlyList<int> AssignClusters(IReadOnlyList<Point> centers, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(centers);
        ArgumentNullException.ThrowIfNull(points);
        return [.. points.Select(point => Enumerable.Range(0, centers.Count)
            .MinBy(k => SquaredDistance(point, centers[k])))];
    }

    /// <summary>クラスタごとに、割り当てられた点の平均を新しい中心にする。点が無ければ前の中心を残す。</summary>
    public static IReadOnlyList<Point> UpdateCenters(
        IReadOnlyList<Point> points, IReadOnlyList<int> labels, IReadOnlyList<Point> previousCenters)
    {
        ArgumentNullException.ThrowIfNull(points);
        ArgumentNullException.ThrowIfNull(labels);
        ArgumentNullException.ThrowIfNull(previousCenters);
        return [.. previousCenters.Select((previous, k) =>
        {
            var members = points.Where((_, i) => labels[i] == k).ToList();
            if (members.Count == 0)
            {
                return previous;
            }

            // コレクション式の前には型変換を書けないので、型を書いた変数に受けてから返す
            Point center = [.. Enumerable.Range(0, previous.Count).Select(j => members.Average(point => point[j]))];
            return center;
        })];
    }

    /// <summary>SSE。各点と、その点が属するクラスタの中心との距離の 2 乗の合計。</summary>
    public static double SumOfSquaredErrors(
        IReadOnlyList<Point> points, IReadOnlyList<int> labels, IReadOnlyList<Point> centers)
    {
        ArgumentNullException.ThrowIfNull(points);
        ArgumentNullException.ThrowIfNull(labels);
        ArgumentNullException.ThrowIfNull(centers);
        return points.Select((point, i) => SquaredDistance(point, centers[labels[i]])).Sum();
    }

    /// <summary>反復回数の上限を既定値にして、初期中心から K-means でクラスタリングする。</summary>
    public static KMeansResult Fit(IReadOnlyList<Point> initialCenters, IReadOnlyList<Point> points) =>
        Fit(MaxIterations, initialCenters, points);

    /// <summary>
    /// 中心が変わらなくなるか、反復回数が上限に達するまで、割り当てと中心の更新を繰り返す。
    /// 初期中心を引数で受け取るので、同じ初期中心を渡せば何度でも同じ結果になる。
    /// </summary>
    public static KMeansResult Fit(int maxIterations, IReadOnlyList<Point> initialCenters, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(initialCenters);
        ArgumentNullException.ThrowIfNull(points);
        var centers = initialCenters;
        for (var i = 0; i < maxIterations; i++)
        {
            var next = UpdateCenters(points, AssignClusters(centers, points), centers);
            var settled = SameCenters(centers, next);
            centers = next;
            if (settled)
            {
                break;
            }
        }

        var labels = AssignClusters(centers, points);
        return new KMeansResult(labels, centers, SumOfSquaredErrors(points, labels, centers));
    }

    /// <summary>点の番号をシード付きで並べ替え、先頭の nClusters 個の点を初期中心にする。</summary>
    public static IReadOnlyList<Point> ChooseInitialCenters(int nClusters, int seed, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(points);
        return [.. Preprocessing.Shuffle([.. Enumerable.Range(0, points.Count)], seed)
            .Take(nClusters)
            .Select(i => points[i])];
    }

    /// <summary>初期中心の候補ごとにクラスタリングし、SSE が最小の結果を返す。</summary>
    public static KMeansResult Best(IReadOnlyList<IReadOnlyList<Point>> candidates, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(candidates);
        return candidates.Select(initialCenters => Fit(initialCenters, points)).MinBy(result => result.Sse)!;
    }

    /// <summary>シードを 1 ずつずらして初期中心を nInit 通り選び、SSE が最小の結果を返す。</summary>
    public static KMeansResult FitWithRestarts(int nInit, int nClusters, int seed, IReadOnlyList<Point> points) =>
        Best(
            [.. Enumerable.Range(0, nInit).Select(i => ChooseInitialCenters(nClusters, seed + i, points))],
            points);

    /// <summary>クラスタ数ごとに、初期中心を nInit 通り試した最小の SSE を求める（エルボー法）。</summary>
    public static IReadOnlyList<KeyValuePair<int, double>> SseByClusterCount(
        int nInit, int seed, IReadOnlyList<int> clusterCounts, IReadOnlyList<Point> points)
    {
        ArgumentNullException.ThrowIfNull(clusterCounts);
        return [.. clusterCounts.Select(n =>
            KeyValuePair.Create(n, FitWithRestarts(nInit, n, seed, points).Sse))];
    }

    private static bool SameCenters(IReadOnlyList<Point> a, IReadOnlyList<Point> b) =>
        a.Zip(b).All(pair => pair.First.SequenceEqual(pair.Second));
}
